### ISSUE TITLE
Automatic update of azure-sdk-for-net with new generator is broken

### ISSUE THREAD

**Issue by ellismg** [Azure/autorest.csharp#1506]
It looks like the automated job to update the code generator version in azure-sdk-for-net has been broken for a few weeks now, but it's not obvious to me from inspection why.

Here's a recent build which failed:

https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1087786&view=logs&j=8d012f1c-7523-5435-2994-2f1676e39128&t=1c4603fe-37d0-5e4a-7ff2-571d26135ecd

From the end of the log:

```
EXEC : warning (PreCheck/CheckDuplicateSchemas): Checking for duplicate schemas, this could take a (long) while.  Run with --verbose for more detail. [/home/vsts/work/1/s/azure-sdk-for-net/sdk/webpubsub/Azure.Messaging.WebPubSub/src/Azure.Messaging.WebPubSub.csproj]
  [13.14 s] Generation Complete
##[error]PowerShell exited with code '1'.
```

I'm not exactly sure why PowerShell is existing with an error code here.  There are tons and tons of lint warnings output by the tool but it seems like there are no errors during code generation.  Running the command:

```
dotnet msbuild /restore /t:GenerateCode .\eng\service.proj
```

Locally after updating works fine, and echoing `$?` from PowerShell after returns `True` so I'm not exactly sure what's going on here.

**Comment by AlexanderSher** [comment]
```
2021-09-09T01:42:04.0388431Z   <--- JS stacktrace --->
2021-09-09T01:42:04.0389768Z   
2021-09-09T01:42:04.0391406Z EXEC : FATAL error : Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory [/home/vsts/work/1/s/azure-sdk-for-net/sdk/network/Azure.ResourceManager.Network/src/Azure.ResourceManager.Network.csproj]
2021-09-09T01:42:04.2144095Z    1: 0xa25510 node::Abort() [/usr/local/bin/node]
2021-09-09T01:42:04.2239657Z    2: 0x9664d3 node::FatalError(char const*, char const*) [/usr/local/bin/node]
2021-09-09T01:42:04.2251873Z    3: 0xb9a8be v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/usr/local/bin/node]
2021-09-09T01:42:04.2261339Z    4: 0xb9ac37 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/usr/local/bin/node]
2021-09-09T01:42:04.2270249Z    5: 0xd56ca5  [/usr/local/bin/node]
2021-09-09T01:42:04.2279495Z    6: 0xd5782f  [/usr/local/bin/node]
2021-09-09T01:42:04.2289537Z    7: 0xd6566b v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/usr/local/bin/node]
2021-09-09T01:42:04.2396603Z    8: 0xd6922c v8::internal::Heap::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/usr/local/bin/node]
2021-09-09T01:42:04.2435285Z    9: 0xd3790b v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationType, v8::internal::AllocationOrigin) [/usr/local/bin/node]
2021-09-09T01:42:04.2441793Z   10: 0x107fbef v8::internal::Runtime_AllocateInYoungGeneration(int, unsigned long*, v8::internal::Isolate*) [/usr/local/bin/node]
2021-09-09T01:42:04.2463077Z   11: 0x1426919  [/usr/local/bin/node]
2021-09-09T01:42:04.6850658Z /home/vsts/.nuget/packages/microsoft.azure.autorest.csharp/3.0.0-beta.20210908.2/buildMultiTargeting/Microsoft.Azure.AutoRest.CSharp.targets(34,5): error MSB3073: The command "npx autorest@3.2.3 --skip-csproj --skip-upgrade-check --version=3.4.5 /home/vsts/work/1/s/azure-sdk-for-net/sdk/network/Azure.ResourceManager.Network/src/autorest.md  --use=/home/vsts/.nuget/packages/microsoft.azure.autorest.csharp/3.0.0-beta.20210908.2/buildMultiTargeting/../tools/netcoreapp3.1/any/ --output-folder=/home/vsts/work/1/s/azure-sdk-for-net/sdk/network/Azure.ResourceManager.Network/src/Generated --clear-output-folder=true --namespace=Azure.ResourceManager.Network --shared-source-folders="/home/vsts/work/1/s/azure-sdk-for-net/eng//../sdk/core/Azure.Core/src/Shared/;/home/vsts/.nuget/packages/microsoft.azure.autorest.csharp/3.0.0-beta.20210908.2/buildMultiTargeting/../content/Generator.Shared/"" exited with code -1. [/home/vsts/work/1/s/azure-sdk-for-net/sdk/network/Azure.ResourceManager.Network/src/Azure.ResourceManager.Network.csproj]
```

**Comment by ellismg** [comment]
Thanks, @AlexanderSher!

It looks like the key difference between the run in CI and running locally on my machine is the maximum amount of memory that's node is allowed to consume.  On the build machines we see the following output:

```
  AutoRest code generation utility [cli version: 3.2.3; node: v14.17.6, max-memory: 2048 MB]
```

and when running locally on my machine, 4GB is assigned:

```
AutoRest code generation utility [cli version: 3.2.1; node: v16.8.0, max-memory: 4096 MB]
```

Looking at Task Manager during the run, it does seem like the process ends up consuming a fair amount of memory.  I was able to reproduce the issue locally by setting the `NODE_OPTIONS` environment variable to `--max-old-space-size=2048` (and noted that when doing so, the `max-memory` value shown by AutoRest when it starts was correctly lowered to 2048.

I then used `NODE_OPTIONS` to set `max-old-space-size` to 3072 (3GB) and the generation was able to complete without error.  Some light searching implies that the AzDo runners have 7GB of memory, so perhaps we can bump this up to 3072 MB without running into tons of swapping on the build machine?

Happy to submit a PR to `eng\UpdateAzureSdkForNet.ps1` if we want to try that out, let me know what you think.
